File: src/ts/domains/card.ts
import { equalStoryPoint, StoryPoint } from "./story-point";

type GiveUpCard = {
  kind: "giveup";
};

type StoryPointCard = {
  kind: "storypoint";
  storyPoint: StoryPoint;
};

export type Card = GiveUpCard | StoryPointCard;

/**
   factory function for GiveUpCard
 */
export const createGiveUpCard = (): GiveUpCard => {
  return {
    kind: "giveup",
  };
};

/**
   factory function for StoryPointCard
 */
export const createStoryPointCard = (storyPoint: StoryPoint): StoryPointCard => {
  return {
    kind: "storypoint",
    storyPoint,
  };
};

export const equalCard = (v1: Card, v2: Card) => {
  if (v1.kind !== v2.kind) {
    return false;
  }

  if (v1.kind === "giveup" && v2.kind === "giveup") {
    return true;
  } else if (v1.kind === "storypoint" && v2.kind === "storypoint") {
    return equalStoryPoint(v1.storyPoint, v2.storyPoint);
  }

  return false;
};

File: src/rust/domains/card.rs
use crate::domains::story_point::StoryPoint;

#[derive(Debug, PartialEq, Clone)]
enum CardInner {
    GiveUpCard,
    StoryPointCard(StoryPoint),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Card(CardInner);

impl Card {
    pub fn giveup() -> Card {
        Card(CardInner::GiveUpCard)
    }

    pub fn storypoint(point: StoryPoint) -> Card {
        Card(CardInner::StoryPointCard(point))
    }

    pub fn as_story_point(&self) -> Option<StoryPoint> {
        match self.0 {
            CardInner::GiveUpCard => None,
            CardInner::StoryPointCard(s) => Some(s),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn should_return_none_if_giveup_card() {
        // arrange
        let v = Card::giveup();

        // do
        let actual = v.as_story_point();

        // arrange
        assert_eq!(actual, None)
    }

    #[test]
    fn should_return_point_if_storypoint_card() {
        // arrange
        let v = Card::storypoint(StoryPoint::new(4));

        // do
        let actual = v.as_story_point();

        // arrange
        assert_eq!(actual, Some(StoryPoint::new(4)))
    }
}
